### PR TITLE
[HiCache] Optimize HiCache host pool free-list release

### PR DESCRIPTION
**Pull request by huangtingwei9988**
## Motivation

HiCache host-pool `free()` currently appends released slots back to `free_slots` with an eager `torch.cat`:

```python
self.free_slots = torch.cat([self.free_slots, indices_cpu])
```

For large host pools, `free_slots` can contain millions of entries. Even when only a small number of slots is released, this operation copies the entire free-list on every `free()` call. In decode, this shows up inside `scheduler.process_batch_result` after HiCache write-back / host eviction and can introduce multi-millisecond CPU spikes.

The fix is to make host-pool slot release lazy, matching the device allocator’s `free_pages + release_pages` pattern:

- `free()` only records released chunks in `release_slots`.
- `available_size()` includes both `free_slots` and pending released slots.
- `alloc()` continues to allocate from `free_slots` directly while possible.
- Only when `free_slots` is insufficient does `alloc()` merge `release_slots` back into `free_slots` once, then proceed with the normal allocation path.

This removes the eager large `torch.cat` from the hot `free()` path while preserving the existing free-list allocation semantics.

## benchmark

In this run, the host KV cache capacity reported by metrics was 6,611,072 token slots 
```
Validation on Ling v2.6, TP=4, batch=1, 2k input / 128 output, HiCache direct/page_first_direct/write_through

Profile result, pre-patch vs current lazy-merge patch:

first scheduler.process_batch_result:
  mean: 5.520 ms -> 2.050 ms
  max:  9.776 ms -> 2.776 ms

aten::cat inside the first process_batch_result:
  mean max: 2.914 ms -> 0.045 ms
  max:      6.207 ms -> 0.052 ms

total scheduler.process_batch_result:
  mean: 24.859 ms -> 12.260 ms
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29303604544](https://github.com/sgl-project/sglang/actions/runs/29303604544)
Latest PR Test (Extra): :white_check_mark: [Run #29303604433](https://github.com/sgl-project/sglang/actions/runs/29303604433)

### CODE REVIEW

**Review by hzh0425 — APPROVED** [review]
LGTM，PLS Add this optimization for MambaPoolHost.free() and DeepSeekV4PagedHostPool.free()

**Comment by hzh0425** [comment]
image
The failed test is unrelated

File: python/sglang/srt/mem_cache/memory_pool_host.py [hunk 1]
```diff
@@ -908,15 +921,22 @@ def alloc(self, need_size: int) -> Optional[torch.Tensor]:
...
         self.free_slots = self.free_slots[need_size:]
         return select_index
 
     @synchronized
     def free(self, indices: torch.Tensor) -> int:
-        self.free_slots = torch.cat(
-            [self.free_slots, indices.to(dtype=torch.int64, device="cpu").flatten()]
-        )
+        indices_cpu = indices.cpu()
+        if indices_cpu.numel() == 0:
```
Comment (hzh0425) [inline python/sglang/srt/mem_cache/memory_pool_host.py:935]: should we still need flatten? @huangtingwei9988

Comment (huangtingwei9988) [reply python/sglang/srt/mem_cache/memory_pool_host.py:935]: `flatten()` is redundant because indices is already 1-D
